## This Jupyter notebook will be used to write all the code used in the Data Science Specialization Capstone Project offered by IBM on Coursera


In [12]:
import pandas as pd
import numpy as np
print('Hello Capstone Project Course')

Hello Capstone Project Course



### 1. This data set was published with an Open Government Licence on  the data.gov.uk website and its titled "Police Recorded Injury Road Traffic Collision Statistics Northern Ireland 2016"

### The dataset for 2017, 2018, and 2019 have been added to this data set to make it richer for analysis

In [13]:
filename = "/home/tunde/Documents/COURSERA-DATA-SCIENCE-SPECIALIZATION/4_APPLIED_DATA_SCIENCE_CAPSTONE/Data_Set/Police_Record/collision2016_to_2019.csv"


In [14]:
df = pd.read_csv(filename, index_col=0)


In [15]:
df.head()

a_ref a_District  a_type  a_veh  a_cas a_wkday  a_day  a_month  \
a_year                                                                   
2016        1       NEMD       3      1      1     FRI      1        1   
2016        2       DCST       3      2      1     FRI      1        1   
2016        3       ARND       2      1      1     FRI      1        1   
2016        4       BELC       3      2      2     FRI      1        1   
2016        5       BELC       3      2      1     FRI      1        1   

        a_hour  a_min  ...  a_jdet  a_jcont  a_pedhum  a_pedphys a_light  \
a_year                 ...                                                 
2016         1     49  ...                                                 
2016         3     19  ...                                                 
2016         3     35  ...       1        1         1          1       6   
2016         3     50  ...                                                 
2016        15      4  ...                                                 

       a_weat a_roadsc a_speccs a_chaz a_scene  
a_year                                          
2016                                            
2016                                            
2016        2        4        1      1       1  
2016                                            
2016                                            

[5 rows x 24 columns]

### 2. Checking the number of entries, columns and data types for each colunm for the data set 

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23731 entries, 2016 to 2019
Data columns (total 24 columns):
a_ref         23731 non-null int64
a_District    23731 non-null object
a_type        23731 non-null int64
a_veh         23731 non-null int64
a_cas         23731 non-null int64
a_wkday       23731 non-null object
a_day         23731 non-null int64
a_month       23731 non-null int64
a_hour        23731 non-null int64
a_min         23731 non-null int64
a_gd1         23731 non-null int64
a_gd2         23731 non-null int64
a_ctype       23731 non-null int64
a_speed       23731 non-null int64
a_jdet        23731 non-null object
a_jcont       23731 non-null object
a_pedhum      23731 non-null object
a_pedphys     23731 non-null object
a_light       23731 non-null object
a_weat        23731 non-null object
a_roadsc      23731 non-null object
a_speccs      23731 non-null object
a_chaz        23731 non-null object
a_scene       23731 non-null object
dtypes: int64(12), object(12)
memory 

In [17]:
df.describe(include = "all")

a_ref a_District       a_type         a_veh         a_cas  \
count   23731.000000      23731  23731.00000  23731.000000  23731.000000   
unique           NaN         11          NaN           NaN           NaN   
top              NaN       BELC          NaN           NaN           NaN   
freq             NaN       5872          NaN           NaN           NaN   
mean     2971.237917        NaN      2.87097      1.867304      1.532721   
std      1720.229069        NaN      0.36335      0.652304      1.018194   
min         1.000000        NaN      1.00000      1.000000      1.000000   
25%      1484.000000        NaN      3.00000      1.000000      1.000000   
50%      2967.000000        NaN      3.00000      2.000000      1.000000   
75%      4450.000000        NaN      3.00000      2.000000      2.000000   
max      6225.000000        NaN      3.00000     10.000000     22.000000   

       a_wkday         a_day       a_month        a_hour         a_min  ...  \
count    23731  23731.000000  23731.000000  23731.000000  23731.000000  ...   
unique       7           NaN           NaN           NaN           NaN  ...   
top        FRI           NaN           NaN           NaN           NaN  ...   
freq      3931           NaN           NaN           NaN           NaN  ...   
mean       NaN     15.652817      6.647929     13.723610     27.327925  ...   
std        NaN      8.765668      3.481886      4.989497     17.733712  ...   
min        NaN      1.000000      1.000000      0.000000      0.000000  ...   
25%        NaN      8.000000      4.000000     10.000000     12.000000  ...   
50%        NaN     16.000000      7.000000     14.000000     29.000000  ...   
75%        NaN     23.000000     10.000000     17.000000     43.000000  ...   
max        NaN     31.000000     12.000000     23.000000     59.000000  ...   

        a_jdet  a_jcont  a_pedhum  a_pedphys a_light a_weat a_roadsc a_speccs  \
count    23731    23731     23731      23731   23731  23731    23731    23731   
unique       9        6         4          7       8     11       11        7   
top                                                                             
freq     20902    20902     20903      20903   20902  20902    20902    20902   
mean       NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   
std        NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   
min        NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   
25%        NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   
50%        NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   
75%        NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   
max        NaN      NaN       NaN        NaN     NaN    NaN      NaN      NaN   

       a_chaz a_scene  
count   23731   23731  
unique      7       3  
top                    
freq    20902   20902  
mean      NaN     NaN  
std       NaN     NaN  
min       NaN     NaN  
25%       NaN     NaN  
50%       NaN     NaN  
75%       NaN     NaN  
max       NaN     NaN  

[11 rows x 24 columns]

### 3. There are a lot of missing values in the last 10 columns. We can confirm that there are 20903 missing values for these columns. This gives us (23731 - 20903) number of entries ( 2828 entries). We may have to drop these columns later

In [18]:
df = df.replace(r'^\s*$', np.nan, regex=True)
df.isnull().sum(axis=0)


a_ref             0
a_District        0
a_type            0
a_veh             0
a_cas             0
a_wkday           0
a_day             0
a_month           0
a_hour            0
a_min             0
a_gd1             0
a_gd2             0
a_ctype           0
a_speed           0
a_jdet        20902
a_jcont       20902
a_pedhum      20903
a_pedphys     20903
a_light       20902
a_weat        20902
a_roadsc      20902
a_speccs      20902
a_chaz        20902
a_scene       20902
dtype: int64

### 4. We have 11 unique Districts for Northern Ireland as entries in the dataset. Here we can see "BELC" (Belfast City) has the highest number of entries at 5872. We might want to observe the how speed limits "a_speed" affects Collision in the Belfast City district 

In [34]:
df['a_District'].value_counts()

BELC    5872
ARBC    2329
NEMD    2114
LISC    2056
ANTN    1887
DCST    1751
MIDU    1684
ARND    1661
CCGL    1641
MEAN    1414
FERO    1322
Name: a_District, dtype: int64

### 5. There are a lot  of missing values in the data as shown in the output below

In [27]:
df.isnull()

a_ref  a_District  a_type  a_veh  a_cas  a_wkday  a_day  a_month  \
a_year                                                                     
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
2016    False       False   False  False  False    False  False    False   
...       ...         ...     ...    ...    ...      ...    ...      ...   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False  False    False   
2019    False       False   False  False  False    False

### 6. In the output below, we see that after dropping all rows that have "NaN" values, we have 754 entries for our training data. 

In [28]:
new_df = df.dropna()

new_df

a_ref a_District  a_type  a_veh  a_cas a_wkday  a_day  a_month  \
a_year                                                                   
2016        3       ARND       2      1      1     FRI      1        1   
2016       10       ANTN       2      3      7     SAT      2        1   
2016       11       ARBC       2      2      2     SAT      2        1   
2016       29       ANTN       2      1      1     SUN      3        1   
2016       35       ARND       1      2      2     MON      4        1   
2016       82       ANTN       2      2      1     WED      6        1   
2016       85       BELC       2      1      3     WED      6        1   
2016      100       NEMD       2      2      2     FRI      8        1   
2016      108       ARBC       2      2      3     FRI      8        1   
2016      126       NEMD       1      3      3     SAT      9        1   
2016      127       FERO       2      2      1     SAT      9        1   
2016      128       DCST       2      2      1     SAT      9        1   
2016      135       NEMD       2      1      1     SAT      9        1   
2016      150       ANTN       2      2      3     MON     11        1   
2016      157       BELC       2      1      1     MON     11        1   
2016      159       ARBC       2      1      1     MON     11        1   
2016      190       MIDU       2      1      2     WED     13        1   
2016      207       MEAN       2      1      1     THU     14        1   
2016      229       CCGL       2      1      1     SAT     16        1   
2016      230       BELC       2      2      2     SAT     16        1   
2016      231       NEMD       2      1      1     SAT     16        1   
2016      241       BELC       2      1      1     SAT     16        1   
2016      249       MEAN       2      2      1     SUN     17        1   
2016      259       ARBC       2      1      1     TUE     19        1   
2016      262       CCGL       2      2      1     TUE     19        1   
2016      286       ARND       1      1      1     WED     20        1   
2016      290       MIDU       2      1      1     WED     20        1   
2016      298       ARND       2      1      1     WED     20        1   
2016      301       CCGL       2      1      1     THU     21        1   
2016      302       ANTN       2      2      1     THU     21        1   
...       ...        ...     ...    ...    ...     ...    ...      ...   
2019     5448       ANTN       1      2      2     TUE     17       12   
2019     5484       ANTN       2      2      2     WED     18       12   
2019     5486       MIDU       1      3      4     WED     18       12   
2019     5494       BELC       2      2      1     THU     19       12   
2019     5503       BELC       2      3      2     THU     19       12   
2019     5514       FERO       2      1      1     THU     19       12   
2019     5520       BELC       2      1      1     FRI     20       12   
2019     5523       ARBC       2      3      2     FRI     20       12   
2019     5528       NEMD       2      3      6     FRI     20       12   
2019     5539       NEMD       1      1      2     SAT     21       12   
2019     5544       MEAN       2      1      1     SAT     21       12   
2019     5553       ARBC       2      1      1     SAT     21       12   
2019     5556       FERO       2      1      1     SAT     21       12   
2019     5562       CCGL       2      1      1     SUN     22       12   
2019     5564       MIDU       2      1      1     SUN     22       12   
2019     5572       ARBC       1      3      1     SUN     22       12   
2019     5589       BELC       2      1      1     SAT     21       12   
2019     5592       MIDU       1      2      3     MON     23       12   
2019     5596       ANTN       2      1      1     TUE     24       12   
2019     5604       ARBC       2      1      1     WED     25       12   
2019     5605       NEMD       2      1      2     THU     26       12   
2019     561

### 7. The number of entries per district has reduced drastically

In [29]:
new_df['a_District'].value_counts()

BELC    463
ARBC    331
NEMD    324
CCGL    249
MIDU    233
LISC    219
ANTN    217
FERO    212
ARND    209
MEAN    194
DCST    177
Name: a_District, dtype: int64

In [30]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2828 entries, 2016 to 2019
Data columns (total 24 columns):
a_ref         2828 non-null int64
a_District    2828 non-null object
a_type        2828 non-null int64
a_veh         2828 non-null int64
a_cas         2828 non-null int64
a_wkday       2828 non-null object
a_day         2828 non-null int64
a_month       2828 non-null int64
a_hour        2828 non-null int64
a_min         2828 non-null int64
a_gd1         2828 non-null int64
a_gd2         2828 non-null int64
a_ctype       2828 non-null int64
a_speed       2828 non-null int64
a_jdet        2828 non-null object
a_jcont       2828 non-null object
a_pedhum      2828 non-null object
a_pedphys     2828 non-null object
a_light       2828 non-null object
a_weat        2828 non-null object
a_roadsc      2828 non-null object
a_speccs      2828 non-null object
a_chaz        2828 non-null object
a_scene       2828 non-null object
dtypes: int64(12), object(12)
memory usage: 552.3+ KB


### 8. Describing the data 

In [31]:
new_df.describe(include = "all")

a_ref a_District       a_type        a_veh        a_cas a_wkday  \
count   2828.000000       2828  2828.000000  2828.000000  2828.000000    2828   
unique          NaN         11          NaN          NaN          NaN       7   
top             NaN       BELC          NaN          NaN          NaN     SAT   
freq            NaN        463          NaN          NaN          NaN     463   
mean    2966.537129        NaN     1.917610     1.665134     1.729491     NaN   
std     1691.323308        NaN     0.275007     0.739956     1.261952     NaN   
min        1.000000        NaN     1.000000     1.000000     1.000000     NaN   
25%     1548.000000        NaN     2.000000     1.000000     1.000000     NaN   
50%     2966.500000        NaN     2.000000     2.000000     1.000000     NaN   
75%     4366.500000        NaN     2.000000     2.000000     2.000000     NaN   
max     6203.000000        NaN     2.000000     7.000000    15.000000     NaN   

              a_day      a_month       a_hour        a_min  ...  a_jdet  \
count   2828.000000  2828.000000  2828.000000  2828.000000  ...    2828   
unique          NaN          NaN          NaN          NaN  ...       8   
top             NaN          NaN          NaN          NaN  ...       1   
freq            NaN          NaN          NaN          NaN  ...    1343   
mean      15.679279     6.639321    13.619165    27.424682  ...     NaN   
std        8.768212     3.422295     5.740800    17.759436  ...     NaN   
min        1.000000     1.000000     0.000000     0.000000  ...     NaN   
25%        8.000000     4.000000    10.000000    12.000000  ...     NaN   
50%       16.000000     7.000000    15.000000    28.000000  ...     NaN   
75%       23.000000    10.000000    18.000000    43.000000  ...     NaN   
max       31.000000    12.000000    23.000000    59.000000  ...     NaN   

        a_jcont  a_pedhum  a_pedphys a_light a_weat a_roadsc a_speccs a_chaz  \
count      2828      2828       2828    2828   2828     2828     2828   2828   
unique        5         3          6       7     10       10        6      6   
top           1         1          1       2      1        1        1      1   
freq       1343      2799       2470     904   1954     1806     2777   2761   
mean        NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   
std         NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   
min         NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   
25%         NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   
50%         NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   
75%         NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   
max         NaN       NaN        NaN     NaN    NaN      NaN      NaN    NaN   

       a_scene  
count     2828  
unique       2  
top          1  
freq      2574  
mean       NaN  
std        NaN  
min        NaN  
25%        NaN  
50%        NaN  
75%        NaN  
max        NaN  

[11 rows x 24 columns]

### 9.  checking how balanced the data is

### I won't be dropping the data. Instead, I will replacing measing values with mode values. As  you can see, dropping all missing values will remove all incidents that resulted in "slight injury(3)" from the dataset

In [32]:
new_df['a_type'].value_counts()

2    2595
1     233
Name: a_type, dtype: int64

In [33]:
df['a_type'].value_counts()

3    20902
2     2596
1      233
Name: a_type, dtype: int64

### 10. A quick look at what the dataset for Belfast City would look like

In [37]:
#df.cats[df.cats=='a']
bel_df = df[df.a_District == 'BELC']
bel_df


a_ref a_District  a_type  a_veh  a_cas a_wkday  a_day  a_month  \
a_year                                                                   
2016        4       BELC       3      2      2     FRI      1        1   
2016        5       BELC       3      2      1     FRI      1        1   
2016        7       BELC       3      2      2     FRI      1        1   
2016        8       BELC       3      2      1     SAT      2        1   
2016        9       BELC       3      2      2     SAT      2        1   
2016       12       BELC       3      2      1     SAT      2        1   
2016       13       BELC       3      2      1     SAT      2        1   
2016       17       BELC       3      2      1     SUN      3        1   
2016       18       BELC       3      2      2     SUN      3        1   
2016       22       BELC       3      2      1     SUN      3        1   
2016       23       BELC       3      2      2     SUN      3        1   
2016       25       BELC       3      2      1     SUN      3        1   
2016       37       BELC       3      2      2     MON      4        1   
2016       42       BELC       3      2      1     MON      4        1   
2016       47       BELC       3      2      2     MON      4        1   
2016       48       BELC       3      3      1     TUE      5        1   
2016       54       BELC       3      2      1     TUE      5        1   
2016       56       BELC       3      2      1     TUE      5        1   
2016       61       BELC       3      2      1     TUE      5        1   
2016       62       BELC       3      2      1     TUE      5        1   
2016       63       BELC       3      2      1     TUE      5        1   
2016       85       BELC       2      1      3     WED      6        1   
2016       89       BELC       3      2      1     THU      7        1   
2016       94       BELC       3      2      1     THU      7        1   
2016      103       BELC       3      2      1     WED      6        1   
2016      107       BELC       3      2      5     FRI      8        1   
2016      111       BELC       3      2      1     FRI      8        1   
2016      118       BELC       3      2      1     FRI      8        1   
2016      124       BELC       3      2      2     FRI      8        1   
2016      131       BELC       3      3      2     SAT      9        1   
...       ...        ...     ...    ...    ...     ...    ...      ...   
2019     5579       BELC       3      2      3     MON     23       12   
2019     5582       BELC       3      2      1     SUN     22       12   
2019     5588       BELC       3      2      3     MON     23       12   
2019     5589       BELC       2      1      1     SAT     21       12   
2019     5599       BELC       3      1      1     TUE     24       12   
2019     5600       BELC       3      2      1     TUE     24       12   
2019     5601       BELC       3      2      1     SAT     21       12   
2019     5608       BELC       3      2      1     THU     26       12   
2019     5610       BELC       3      2      1     THU     26       12   
2019     5613       BELC       3      2      1     THU     26       12   
2019     5614       BELC       3      2      2     THU     26       12   
2019     5616       BELC       3      2      2     THU     26       12   
2019     5617       BELC       1      1      1     THU     26       12   
2019     5620       BELC       3      1      1     THU     26       12   
2019     5625       BELC       3      2      1     FRI     27       12   
2019     5626       BELC       3      2      1     FRI     27       12   
2019     5628       BELC       3      2      3     FRI     27       12   
2019     5632       BELC       2      1      1     FRI     27       12   
2019     5633       BELC       3      1      1     MON     23       12   
2019     5642       BELC       3      1      3     SAT     28       12   
2019     5652       BELC       3      1      1     SUN     29       12   
2019     565

## 11. Checking how balanced the dataset for Belfast City is 

### observations for collison severity are still unbalanced

In [38]:
bel_df['a_type'].value_counts()

3    5409
2     449
1      14
Name: a_type, dtype: int64